### PR TITLE
Add an image that includes Intel's toolchain

### PR DESCRIPTION
**Pull request by renefritze**

### CODE REVIEW

**Comment by renefritze** [comment]
This follows https://software.intel.com/content/www/us/en/develop/articles/oneapi-repo-instructions.html#aptpkg
but I haven't tested this yet since the even the basekit installs like 14GB worth of packages :open_mouth: 

Maybe @tobiasleibner is interested though.

**Comment by renefritze** [comment]
Was this usable as-is, @tobiasleibner ?

**Comment by tobiasleibner** [comment]
It was a good start, but I had to do some additional changes to make it work (see the ``full-intel`` file in my recent commits). Changes I remember are:
- also install intel-hpckit
- do not install openmpi to avoid conflicts
- do not install metis to avoid conflicts
- install some packages (e.g. eigen, lpsolve, suitesparse) after installing the intel toolkits to make them find the intel mpi
- update qhull to fix compilation error
I still have to run
```
source /opt/intel/oneapi/setvars.sh -i_mpi_library_kind=release
```
manually, this should probably also be included in the Dockerfile. The default MPI library kind is release_mt, but this gives the warning
```
release_mt library was used but no multi-ep feature was enabled. please use release library instead.
```
on every run of an executable that uses MPI. Not sure if we should rather enable this mysterious multi-ep feature instead of using the release library. With these changes, everything seems to work fine, except that I am still getting the warning
```
CMake Warning at /home/tobias/dune-gdt-super-icc/dune-common/cmake/modules/DuneTestMacros.cmake:336 (add_executable):
  Cannot generate a safe runtime search path for target
  test_generic_function__3d_simplex_alunonconformgrid because files in some
  directories may conflict with libraries in implicit directories:

    runtime library [libtbbmalloc.so.2] in /opt/intel/oneapi/tbb/2021.1-beta08/lib/intel64/gcc4.8 may be hidden by files in:
      /opt/intel/oneapi/intelpython/latest/lib
    runtime library [libtbb.so.12] in /opt/intel/oneapi/tbb/2021.1-beta08/lib/intel64/gcc4.8 may be hidden by files in:
      /opt/intel/oneapi/intelpython/latest/lib
```
for each added executable during the CMake run. Not sure what to do about this, maybe remove  ``/opt/intel/oneapi/intelpython/latest/lib`` manually from the ``LD_LIBRARY_PATH`` after sourcing the ``set_vars.sh`` file?

**Comment by renefritze** [comment]
> * update qhull to fix compilation error

That prolly can be changed elsewhere too.

>   I still have to run
> 
> ```
> source /opt/intel/oneapi/setvars.sh -i_mpi_library_kind=release
> ```

I'll try to put that in the global env then.

> on every run of an executable that uses MPI. Not sure if we should rather enable this mysterious multi-ep feature instead of using the release library. With these changes, everything seems to work fine, except that I am still getting the warning
> 
Indeed we do not need the multi-ep feature since applications have to modified to support them AFAIK.

> ```
> CMake Warning at /home/tobias/dune-gdt-super-icc/dune-common/cmake/modules/DuneTestMacros.cmake:336 (add_executable):
>   Cannot generate a safe runtime search path for target
>   test_generic_function__3d_simplex_alunonconformgrid because files in some
>   directories may conflict with libraries in implicit directories:
> 
>     runtime library [libtbbmalloc.so.2] in /opt/intel/oneapi/tbb/2021.1-beta08/lib/intel64/gcc4.8 may be hidden by files in:
>       /opt/intel/oneapi/intelpython/latest/lib
>     runtime library [libtbb.so.12] in /opt/intel/oneapi/tbb/2021.1-beta08/lib/intel64/gcc4.8 may be hidden by files in:
>       /opt/intel/oneapi/intelpython/latest/lib
> ```
> 
> for each added executable during the CMake run. Not sure what to do about this, maybe remove `/opt/intel/oneapi/intelpython/latest/lib` manually from the `LD_LIBRARY_PATH` after sourcing the `set_vars.sh` file?

Weird. Is that a symlinked dir?

**Comment by tobiasleibner** [comment]
Yes, the TBB libraries in `/opt/intel/oneapi/intelpython/latest/lib` are symlinks to the libraries in `/opt/intel/oneapi/tbb/2021.1-beta08/lib/intel64/gcc4.8`, and both paths appear in the `LD_LIBRARY_PATH` after sourcing the `set_vars.sh` file.

**Comment by tobiasleibner** [comment]
As far as I can see, qhull only appears in `debian/include/full` and in `debian/include/full-intel`, and both are updated to use the 2020 version already in this PR.

**Comment by renefritze** [comment]
I re-ran the whole azure pipeline and it succeeds now. Should this be merged now?

**Comment by tobiasleibner** [comment]
> I'll try to put that in the global env then.

Did you already do that? Other than that, this can be merged :+1:.

**Comment by renefritze** [comment]
> > I'll try to put that in the global env then.
> 
> Did you already do that? Other than that, this can be merged .

Ah, no. I did not.

**Comment by renefritze** [comment]
> > > I'll try to put that in the global env then.
> > 
> > 
> > Did you already do that? Other than that, this can be merged .
> 
> Ah, no. I did not.

Now I did. @tobiasleibner could you check the new image once the pipeline is through?

**Comment by tobiasleibner** [comment]
Yes, I will test the image once the pipeline is green. I just pushed another fix, for some reason the `CLANG_TOOLING_VERSION` variable was not evaluated.

**Comment by renefritze** [comment]
I think that's just poor nomenclature on my part. The tooling version is a [m4 define](https://github.com/dune-community/Dockerfiles/pull/82/files#diff-0d2c7c6118bcf0491dba3a2750fb5caadc1a5fec440f9bf979f762cccfa3068fR3)

**Comment by tobiasleibner** [comment]
Oh, I did not know m4 defines before. Still, my fix for the pipeline seems to work...

**Comment by renefritze** [comment]
Maybe `apt install -qy ` ignores missing packages?

**Comment by tobiasleibner** [comment]
Nope, the [previous pipeline](https://dev.azure.com/dune-community/dockerfiles/_build/results?buildId=1510&view=logs&jobId=41f2a3f9-e2c2-50d2-1126-0dc859a1adb9&j=41f2a3f9-e2c2-50d2-1126-0dc859a1adb9&t=03f283af-64c1-5d2b-eca4-80efcbf2d65d) failed with error message ``Unable to locate package clang-tidy-CLANG_TOOLING_VERSION``. After replacing ``clang-tidy-CLANG_TOOLING_VERSION`` with ``clang-tidy-${CLANG_TOOLING_VERSION}``, it seems to work. So apparently that m4 define is converted to a bash variable somewhere (?).

**Comment by renefritze** [comment]
This is indeed very strange. I'm do now know where the `ENV` setting thing should happen. But if this works here, we'll merge and I'll worry about that later while I prefix all the m4 defines so we can clearly tell them apart from env ones.

**Comment by tobiasleibner** [comment]
On a second thought, I am not sure which image to test, as the Intel images are not build in CI as far as I can tell. Should we add them?

Locally I also see what happens with the ``CLANG_TOOLING_VERSION``. It is not defined in ``debian/include/full`` but only in ``debian/include/full-intel`` which results in the ``E: Unable to locate package clang-tidy-CLANG_TOOLING_VERSION`` error when building the ``debian-full`` image. When replacing ``clang-tidy-CLANG_TOOLING_VERSION`` with ``clang-tidy-${CLANG_TOOLING_VERSION}``, the failure vanishes. However, I can't find ``clang-format`` or ``clang-format-8`` in the resulting image, so it seems to fail silently in that case (?). 
So the correct fix seems to be to add the ``CLANG_TOOLING_VERSION`` define to ``debian/include/full`` instead of adding ``${..}``. But we should really clean up afterwards, grepping for ``CLANG_TOOLING_VERSION`` shows that we have a mixture of ``${CLANG_TOOLING_VERSION}`` and ``CLANG_TOOLING_VERSION``

By the way, the two files ``full`` and ``full-intel`` both have ``clang-tidy-${CLANG_VERSION} clang-tidy-CLANG_TOOLING_VERSION``, is that duplication intended?

**Comment by tobiasleibner** [comment]
Unfortunately, adding the intel lib path to ``LD_LIBRARY_PATH`` is not enough, I still need to run
```
source /opt/intel/oneapi/setvars.sh -i_mpi_library_kind=release
```
before running dunecontrol. Otherwise, it says
```
CMake Error at CMakeLists.txt:2 (project):
  The CMAKE_C_COMPILER:

    icc

  is not a full path and was not found in the PATH.
```
So we probably at least need to add ``/opt/intel/oneapi/compiler/2021.1-beta09/linux/bin/intel64`` (or something similar) to the path. Or can we simply run the ``source`` command above?

**Comment by renefritze** [comment]
Running the source command only has an effect for the current `RUN` directive. I would suggest you do this:

```
printenv > pre
source thescript
printenv > post
diff pre post
```
and add that to the dockerfile.

Another option would be to include the sourcing in the startup files for the shells, but in my experience that's somewhat brittle.

File: debian/include/full [hunk 1]
```diff
@@ -1,7 +1,7 @@
 include(global_macros)
 include(minimal)
 
-define(`CLANG_TOOLING_VERSION', `8')
+ifelse(DEBIANVERSION, `buster', `define(`CLANG_TOOLING_VERSION', `8')', `define(`CLANG_TOOLING_VERSION', `10')')
```
Comment (renefritze) [inline debian/include/full]: I've fixed missing packages in master, maybe rebase this PR instead?

Comment (tobiasleibner) [reply debian/include/full]: I reset this branch to master and started again, only including the relevant changes.

**Comment by tobiasleibner** [comment]
> Running the source command only has an effect for the current `RUN` directive. I would suggest you do this:
> 
> ```
> printenv > pre
> source thescript
> printenv > post
> diff pre post
> ```
> 
> and add that to the dockerfile.

Did that, this gave a lot of environment variables, but it works for now, at least locally. I guess we just have to wait and see if this breaks with every new Intel toolkit version or if these defines are somewhat stable.

**Pull request merged**